qrest/exception.py

```diff
@@ -78,62 +77,71 @@ class RestResourceMissingContentError(RestClientResourceError):
     pass
 
 
-class RestResourceNotFoundError(RestClientResourceError):
+class RestCredentailsError(RestClientResourceError):
     """ wrapper exception """
 
     pass
 
 
-class RestAccessDeniedError(RestClientResourceError):
-    """ wrapper exception """
+class RestResponseError(RestClientResourceError):
+    """Exception due to a requests.Response whose status code indicates an error.
 
-    pass
+    :ivar response: Response object returned by the request
 
+    """
 
-class RestCredentailsError(RestClientResourceError):
-    """ wrapper exception """
+    response: Response
 
-    pass
+    def __init__(self, response: Response, message: str):
+        super().__init__(message)
+        self.response = response
 
 
-class RestInternalServerError(RestClientResourceError):
-    """ wrapper exception """
+class RestAccessDeniedError(RestResponseError):
+    def __init__(self, response: Response):
+        super().__init__(
+            response, f"error {response.status_code}: Access is denied to resource {response.url}"
+        )
 
-    pass
 
+class RestBadRequestError(RestResponseError):
+    def __init__(self, response: Response):
+        super().__init__(response, f"Bad request for resource {response.url}")
 
-class RestBadRequestError(RestClientResourceError):
-    """ wrapper excpetion """
 
-    pass
+class RestInternalServerError(RestResponseError):
+    def __init__(self, response: Response):
+        super().__init__(
+            response, f"error {response.status_code}: Internal Server error ({response.reason})"
+        )
 
 
-class RestResourceHTTPError(HTTPError):
-    """An error when specifying an invalid target for a given REST API."""
+class RestResourceNotFoundError(RestResponseError):
+    def __init__(self, response: Response):
+        super().__init__(response, "Object could not be found in database")
```

Comment:
This results from the 404 response, for which it would be useful for the user to know what the exact error message is. I propose to include the reason in the exception message for this one.